src/Components/TestExplorer.fs

```diff
@@ -1469,10 +1735,208 @@ module Interactions =
+                        rootTestCollection.replace (ResizeArray discoveredTestsAccumulator)
+                    with e ->
+                        logger.Debug("Incremental test discovery update threw an exception", e)
+
+                report "Discovering tests"
+                let! discoveryResponse = LanguageService.discoverTests onTestDiscoveryProgress ()
+
+                let testItems =
+                    discoveryResponse.Data
+                    |> TestItem.ofTestDTOs testItemFactory tryGetLocation
+                    |> ResizeArray
+
+                rootTestCollection.replace (testItems)
+
+                if testItems |> Seq.length = 0 then
+                    window.showWarningMessage (
+                        $"No tests discovered. Make sure your projects are restored, built, and can be run with dotnet test. Discovery logs can be found in Output > F# - Test Adapter "
+                    )
+                    |> ignore
+            }
+
+    let private runTests_WithLanguageServer
+        mergeTestResultsToExplorer
+        (rootTestCollection: TestItemCollection)
+        (req: TestRunRequest)
+        testRun
+        =
+        promise {
+            try
+                let expectedToRun =
+                    req.``include``
+                    |> Option.map Array.ofSeq
+                    |> Option.defaultValue (rootTestCollection.TestItems())
+                    |> Array.collect TestItem.runnableChildren
+
+                let expectedTestsById = expectedToRun |> Array.map (fun t -> t.id, t) |> Map
+
+                let mergeResults (shouldTrim: TrimMissing) (resultDtos: TestResultDTO array) =
+                    let actuallyRan: TestResult array =
+                        resultDtos |> Array.map TestResult.ofTestResultDTO
+
+                    mergeTestResultsToExplorer testRun shouldTrim expectedToRun actuallyRan
+
+                let showStarted (testItems: TestItemDTO array) =
+                    try
+                        let groups = testItems |> Array.groupBy (fun t -> t.ProjectFilePath)
+
+                        groups
+                        |> Array.iter (fun (projPath, activeTests) ->
+                            let testIdsToStart =
+                                activeTests |> Array.map (fun t -> TestItem.constructId projPath t.FullName)
+
+                            let knownExplorerItems = testIdsToStart |> Array.choose expectedTestsById.TryFind
+                            knownExplorerItems |> TestRun.showStarted testRun)
+                    with ex ->
+                        logger.Debug("Threw error while mapping active test items to the explorer", ex)
+
+                let onTestRunProgress (progress: TestRunProgress) =
+                    showStarted progress.ActiveTests
+                    mergeResults TrimMissing.NoTrim progress.TestResults
+
+                    let appendTestResultToOutput (testResult: TestResultDTO) =
```

Comment:
I found printing all the test results to the test output was helpful feedback. I felt more confident that the test run went as expected or could see any error messages.

I could see an argument that printing all the test results feels noisy for big test runs.